### PR TITLE
ARROW-2328: [C++] Fixed and unit tested feather writing with slice

### PR DESCRIPTION
**Pull request by Adriandorr**
When writing a slice of a table to feather, both the scalar data and the null bitmap were written out wrongly, i.e. from the beginning of the original table instead of from the offset.
- scalar data is written from the offset
- null bitmap is written from the offset.

I messed up my original pull request https://github.com/apache/arrow/pull/1766, hence this replaces it. Hope that helps.

### CODE REVIEW

**Comment by wesm** [comment]
BTW, feel no need to put these "ARROW-2328: [C++]" messages in your commits. The only one that matters is in the PR title

**Comment by Adriandorr** [comment]
Will do, thanks.

File: cpp/src/arrow/ipc/test-common.h [hunk 1]
```diff
@@ -223,15 +223,17 @@ Status MakeRandomBinaryArray(int64_t length, bool include_nulls, MemoryPool* poo
...
+          i < int64_t(values.size()) ? values[values_index] : std::to_string(i);
       RETURN_NOT_OK(builder.Append(reinterpret_cast<const RawType*>(value.data()),
                                    static_cast<int32_t>(value.size())));
     }
   }
   return builder.Finish(out);
 }
 
-Status MakeStringTypesRecordBatch(std::shared_ptr<RecordBatch>* out) {
+Status MakeStringTypesRecordBatchWithNulls(std::shared_ptr<RecordBatch>* out,
```
Comment (pitrou) [inline cpp/src/arrow/ipc/test-common.h]: Style note: if it's called `MakeStringTypesRecordBatchWithNulls`, it shouldn't take an argument indicating whether you want to include nulls. Consider renaming or using a private helper function.

File: cpp/src/arrow/ipc/feather.cc [hunk 2]
```diff
@@ -558,15 +595,18 @@ class TableWriter::TableWriterImpl : public ArrayVisitor {
       // byte boundary, and we write this much data into the stream
```
Comment (pitrou) [inline cpp/src/arrow/ipc/feather.cc]: This comment needs updating.

File: cpp/src/arrow/ipc/test-common.h (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/ipc/test-common.h]: Or simply add a `bool with_nulls = true` optional argument to `MakeStringTypesRecordBatch`.

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for doing this. I added a couple review comments, please read.

File: cpp/src/arrow/ipc/test-common.h (hunk 1 above)
Comment (Adriandorr) [reply cpp/src/arrow/ipc/test-common.h]: Adding the default argument didn't work, because MakeStringTypesRecordBatch is used as a function pointer in other tests. I'll rename is to MakeStringTypesRecordBatchWithoutNulls abd this one to MakeStringTypesRecordBatch.

File: cpp/src/arrow/ipc/feather-test.cc [hunk 3]
```diff
@@ -406,6 +406,125 @@ TEST_F(TestTableWriter, PrimitiveNullRoundTrip) {
...
+  ASSERT_OK(MakeStringTypesRecordBatchWithNulls(&batch, false));
+  batch = batch->Slice(320, 30);
+
+  ASSERT_OK(writer_->Append("f0", *batch->column(0)));
+  ASSERT_OK(writer_->Append("f1", *batch->column(1)));
+  Finish();
+
+  std::shared_ptr<Column> col;
+  ASSERT_OK(reader_->GetColumn(0, &col));
+  SCOPED_TRACE(col->data()->chunk(0)->ToString() + "\n" + batch->column(0)->ToString());
```
[the remark this answers is not in the document]
Comment (Adriandorr) [reply cpp/src/arrow/ipc/feather-test.cc]: Yes(ish) It helps give a helpfull message when the test fails.

File: cpp/src/arrow/ipc/feather-test.cc [hunk 4]
```diff
@@ -406,6 +406,125 @@ TEST_F(TestTableWriter, PrimitiveNullRoundTrip) {
...
+  ASSERT_TRUE(col->data()->chunk(0)->Equals(batch->column(0)));
+  ASSERT_EQ("f0", col->name());
+
+  ASSERT_OK(reader_->GetColumn(1, &col));
+  ASSERT_TRUE(col->data()->chunk(0)->Equals(batch->column(1)));
+  ASSERT_EQ("f1", col->name());
+}
+
+TEST_F(TestTableWriter, SliceAtNonEightOffsetStringsWithNullsRoundTrip) {
```
[the remark this answers is not in the document]
Comment (Adriandorr) [reply cpp/src/arrow/ipc/feather-test.cc]: I think the problem with loops in unit tests is that you only see the first failure. That said I should be able to convert these tests to one parameterised test, which hopefully also answers the above concerns..

File: cpp/src/arrow/ipc/feather.cc [hunk 5]
```diff
@@ -75,6 +75,43 @@ static Status WritePadded(io::OutputStream* stream, const uint8_t* data, int64_t
   return Status::OK();
 }
 
+static Status WritePaddedWithOffset(io::OutputStream* stream, const uint8_t* data,
+                                    int64_t bit_offset, const int64_t length,
+                                    int64_t* bytes_written) {
+  data = data + bit_offset / 8;
+  uint8_t bit_shift = static_cast<uint8_t>(bit_offset % 8);
```
[the remark this answers is not in the document]
Comment (Adriandorr) [reply cpp/src/arrow/ipc/feather.cc]: Not really, but I think I got some warning, maybe with clang.

File: cpp/src/arrow/ipc/feather.cc (hunk 2 above)
Comment (Adriandorr) [reply cpp/src/arrow/ipc/feather.cc]: will do.

File: cpp/src/arrow/ipc/feather-test.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/ipc/feather-test.cc]: Cool :-) Though I don't think using a loop is a big deal either...

**Comment by Adriandorr** [comment]
Thanks for the review comments. I’ve rewritten the tests as parameter unit tests and did most of the rest.

File: cpp/src/arrow/ipc/feather-test.cc [hunk 6]
```diff
@@ -406,6 +406,89 @@ TEST_F(TestTableWriter, PrimitiveNullRoundTrip) {
...
+  ASSERT_EQ("f0", col->name());
+
+  ASSERT_OK(reader_->GetColumn(1, &col));
+  ASSERT_TRUE(col->data()->chunk(0)->Equals(batch->column(1)));
+  ASSERT_EQ("f1", col->name());
+}
+
+INSTANTIATE_TEST_CASE_P(
+    TestTableWriterSliceOffsets, TestTableWriterSlice,
+    ::testing::Values(std::make_tuple(300, 30), std::make_tuple(301, 30),
```
Comment (pitrou) [inline cpp/src/arrow/ipc/feather-test.cc]: Hmm... Does gtest allow making a Cartesian product here?

File: cpp/src/arrow/ipc/feather-test.cc [hunk 7]
```diff
@@ -406,6 +406,89 @@ TEST_F(TestTableWriter, PrimitiveNullRoundTrip) {
...
+TEST_P(TestTableWriterSlice, SliceRoundTrip) {
+  auto p = GetParam();
+  auto start = std::get<0>(p);
+  auto size = std::get<1>(p);
+
+  std::shared_ptr<RecordBatch> batch;
+  ASSERT_OK(MakeIntBatchSized(start * 2, &batch));
+  batch = batch->Slice(start, size);
+
+  ASSERT_OK(writer_->Append("f0", *batch->column(0)));
```
Comment (pitrou) [inline cpp/src/arrow/ipc/feather-test.cc]: This part and below is still duplicated in the other two test cases. Do you think you can factor that out, for example as a template function?

File: cpp/src/arrow/ipc/test-common.h [hunk 8]
```diff
@@ -244,18 +246,22 @@ Status MakeStringTypesRecordBatch(std::shared_ptr<RecordBatch>* out) {
...
   {
-    auto s = MakeRandomBinaryArray<BinaryBuilder, uint8_t>(length, true, pool, &a1);
+    auto s = MakeRandomBinaryArray<BinaryBuilder, uint8_t>(length, with_nulls, pool, &a1);
     RETURN_NOT_OK(s);
   }
   *out = RecordBatch::Make(schema, length, {a0, a1});
   return Status::OK();
 }
 
+Status MakeStringTypesRecordBatchWithoutNulls(std::shared_ptr<RecordBatch>* out) {
```
Comment (pitrou) [inline cpp/src/arrow/ipc/test-common.h]: Why is it called "...WithoutNulls" if it passes `with_nulls = true`?

File: cpp/src/arrow/ipc/feather-test.cc (hunk 7 above)
Comment (Adriandorr) [reply cpp/src/arrow/ipc/feather-test.cc]: You are right the duplication can be removed without adding any complexity.

File: cpp/src/arrow/ipc/test-common.h (hunk 8 above)
Comment (Adriandorr) [reply cpp/src/arrow/ipc/test-common.h]: Oops.. I'll rename it.

File: cpp/src/arrow/ipc/feather-test.cc (hunk 6 above)
Comment (Adriandorr) [reply cpp/src/arrow/ipc/feather-test.cc]: Yes it does, yesterday I was just too lazy to find out how.

File: cpp/src/arrow/ipc/test-common.h [hunk 9]
```diff
@@ -223,15 +223,17 @@ Status MakeRandomBinaryArray(int64_t length, bool include_nulls, MemoryPool* poo
     if (include_nulls && values_index == 0) {
       RETURN_NOT_OK(builder.AppendNull());
     } else {
-      const std::string& value = values[values_index];
+      const std::string value =
+          i < int64_t(values.size()) ? values[values_index] : std::to_string(i);
```
Comment (pitrou) [inline cpp/src/arrow/ipc/test-common.h]: Why did you need this change? Is this just a debugging leftover?

Comment (Adriandorr) [reply cpp/src/arrow/ipc/test-common.h]: Not knowing the history of this particular function, I don't know what would be "better" in there. For my test I pretty much just want the consecutive numbers, otherwise it is very difficult to see what has gone wrong (I like the tests to give me the answer to that question if possible). I made a change to add a second function that implements that, but turns out this function is only used from MakeStringTypesRecordBatch, so we can't really have two implementations.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
Thank you! I will merge once the AppVeyor build passes (the Travis-CI failures in the Rust and glib builds are unrelated).
